File: src/main/java/com/enterprise/iqk/evaluation/EvaluationService.java
package com.enterprise.iqk.evaluation;

import com.enterprise.iqk.evaluation.vo.EvalCaseCreateVO;
import com.enterprise.iqk.evaluation.vo.EvalComparisonVO;
import com.enterprise.iqk.evaluation.vo.EvalDatasetCreateVO;
import com.enterprise.iqk.evaluation.vo.EvalDatasetVO;
import com.enterprise.iqk.evaluation.vo.EvalMetricSummaryVO;
import com.enterprise.iqk.evaluation.vo.EvalResultVO;
import com.enterprise.iqk.evaluation.vo.EvalRunRequestVO;
import com.enterprise.iqk.evaluation.vo.EvalRunVO;
import com.enterprise.iqk.rag.HybridRagAnswerService;
import com.enterprise.iqk.retrieval.CitationItem;
import com.enterprise.iqk.retrieval.EvidenceItem;
import com.enterprise.iqk.security.TenantContext;
import com.enterprise.iqk.util.ConversationIdHelper;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

@Service
@RequiredArgsConstructor
public class EvaluationService {
    private static final double PASS_THRESHOLD = 0.70;

    private final EvalDatasetMapper evalDatasetMapper;
    private final EvalCaseMapper evalCaseMapper;
    private final EvalRunMapper evalRunMapper;
    private final EvalResultMapper evalResultMapper;
    private final HybridRagAnswerService hybridRagAnswerService;
    private final ObjectMapper objectMapper;
    private final EvaluationScorer evaluationScorer;
    private final EvaluationReportRenderer evaluationReportRenderer;

    public EvalDatasetVO createDataset(String tenantId, EvalDatasetCreateVO request) {
        if (request == null) {
            throw new IllegalArgumentException("dataset payload is required");
        }
        if (!StringUtils.hasText(request.getName())) {
            throw new IllegalArgumentException("dataset name is required");
        }
        if (request.getCases() == null || request.getCases().isEmpty()) {
            throw new IllegalArgumentException("dataset cases are required");
        }

        String tenant = TenantContext.normalize(tenantId);
        String datasetId = "eval-ds-" + shortUuid();
        LocalDateTime now = LocalDateTime.now();
        EvalDatasetRecord dataset = EvalDatasetRecord.builder()
                .datasetId(datasetId)
                .tenantId(tenant)
                .name(request.getName().trim())
                .description(emptyIfBlank(request.getDescription()))
                .createdAt(now)
                .updatedAt(now)
                .build();
        evalDatasetMapper.insert(dataset);

        int order = 0;
        for (EvalCaseCreateVO item : request.getCases()) {
            if (item == null || !StringUtils.hasText(item.getQuestion())) {
                throw new IllegalArgumentException("case question is required");
            }
            String caseId = StringUtils.hasText(item.getCaseId())
                    ? item.getCaseId().trim()
                    : "case-" + String.format(Locale.ROOT, "%03d", order + 1);
            evalCaseMapper.insert(EvalCaseRecord.builder()
                    .caseId(caseId)
                    .datasetId(datasetId)
                    .tenantId(tenant)
                    .category(emptyIfBlank(item.getCategory()))
                    .chatId(emptyIfBlank(item.getChatId()))
                    .questionText(item.getQuestion().trim())
                    .expectedCitationsJson(writeJsonList(item.getExpectedCitations()))
                    .expectedKeywordsJson(writeJsonList(item.getExpectedKeywords()))
                    .forbiddenKeywordsJson(writeJsonList(item.getForbiddenKeywords()))
                    .sortOrder(order++)
                    .createdAt(now)
                    .updatedAt(now)
                    .build());
        }
        return toDatasetVO(dataset, order);
    }

    public List<EvalDatasetVO> listDatasets(String tenantId) {
        String tenant = TenantContext.normalize(tenantId);
        return evalDatasetMapper.findByTenant(tenant).stream()
                .map(record -> toDatasetVO(record, evalCaseMapper.findByTenantAndDatasetId(tenant, record.getDatasetId()).size()))
                .toList();
    }

    public EvalRunVO triggerRun(String tenantId, String datasetId, EvalRunRequestVO request) {
        String tenant = TenantContext.normalize(tenantId);
        EvalDatasetRecord dataset = requireDataset(tenant, datasetId);
        List<EvalCaseRecord> cases = evalCaseMapper.findByTenantAndDatasetId(tenant, datasetId);
        if (cases.isEmpty()) {
            throw new IllegalArgumentException("dataset has no cases");
        }

        String runId = "eval-run-" + shortUuid();
        LocalDateTime now = LocalDateTime.now();
        String modelProfile = request == null || !StringUtils.hasText(request.getModelProfile())
                ? "balanced"
                : request.getModelProfile().trim();
        EvalRunRecord run = EvalRunRecord.builder()
                .runId(runId)
                .datasetId(dataset.getDatasetId())
                .tenantId(tenant)
                .status("RUNNING")
                .modelProfile(modelProfile)
                .totalCases(cases.size())
                .passedCases(0)
                .runScore(0.0)
                .retrievalHitRate(0.0)
                .citationCoverageRate(0.0)
                .answerFaithfulnessScore(0.0)
                .avgLatencyMs(0.0)
                .failureRate(0.0)
                .startedAt(now)
                .createdAt(now)
                .updatedAt(now)
                .build();
        evalRunMapper.insert(run);

        List<EvalResultRecord> results = new ArrayList<>();
        for (int i = 0; i < cases.size(); i++) {
            results.add(runCase(tenant, runId, dataset.getDatasetId(), cases.get(i), request, i));
        }

        EvalMetricSummaryVO summary = summarize(results);
        run.setStatus("SUCCESS");
        run.setPassedCases(summary.getPassedCases());
        run.setRunScore(summary.getRunScore());
        run.setRetrievalHitRate(summary.getRetrievalHitRate());
        run.setCitationCoverageRate(summary.getCitationCoverageRate());
        run.setAnswerFaithfulnessScore(summary.getAnswerFaithfulnessScore());
        run.setAvgLatencyMs(summary.getAvgLatencyMs());
        run.setFailureRate(summary.getFailureRate());
        run.setFinishedAt(LocalDateTime.now());
        run.setUpdatedAt(run.getFinishedAt());
        evalRunMapper.updateById(run);

        return toRunVO(run, results);
    }

    public EvalRunVO getRun(String tenantId, String runId) {
        String tenant = TenantContext.normalize(tenantId);
        EvalRunRecord run = requireRun(tenant, runId);
        return toRunVO(run, evalResultMapper.findByTenantAndRunId(tenant, runId));
    }

    public EvalRunVO markBaseline(String tenantId, String runId) {
        String tenant = TenantContext.normalize(tenantId);
        EvalRunRecord run = requireRun(tenant, runId);
        int updated = evalDatasetMapper.updateBaselineRunId(tenant, run.getDatasetId(), runId);
        if (updated <= 0) {
            throw new IllegalArgumentException("dataset not found");
        }
        return getRun(tenant, runId);
    }

    public EvalComparisonVO compareLatest(String tenantId, String datasetId) {
        String tenant = TenantContext.normalize(tenantId);
        EvalDatasetRecord dataset = requireDataset(tenant, datasetId);
        List<EvalRunRecord> recent = evalRunMapper.findRecentByDatasetId(tenant, datasetId, 2);
        EvalRunVO current = recent.isEmpty()
                ? null
                : toRunVO(recent.get(0), evalResultMapper.findByTenantAndRunId(tenant, recent.get(0).getRunId()));

        EvalRunVO baseline = null;
        if (StringUtils.hasText(dataset.getBaselineRunId())) {
            EvalRunRecord baselineRun = evalRunMapper.findByTenantAndRunId(tenant, dataset.getBaselineRunId());
            if (baselineRun != null) {
                baseline = toRunVO(baselineRun, evalResultMapper.findByTenantAndRunId(tenant, baselineRun.getRunId()));
            }
        }
        if (baseline == null && recent.size() > 1) {
            EvalRunRecord previous = recent.get(1);
            baseline = toRunVO(previous, evalResultMapper.findByTenantAndRunId(tenant, previous.getRunId()));
        }

        return EvalComparisonVO.builder()
                .dataset(toDatasetVO(dataset, evalCaseMapper.findByTenantAndDatasetId(tenant, datasetId).size()))
                .baseline(baseline)
                .current(current)
                .build();
    }

    public String exportReport(String tenantId, String runId) {
        return evaluationReportRenderer.render(getRun(tenantId, runId));
    }

    private EvalResultRecord runCase(String tenant,
                                     String runId,
                                     String datasetId,
                                     EvalCaseRecord evalCase,
                                     EvalRunRequestVO request,
                                     int index) {
        long startedNs = System.nanoTime();
        String status = "SUCCESS";
        String answer = "";
        String errorMessage = "";
        List<String> citations = List.of();
        List<String> evidence = List.of();

        try {
            String chatId = resolveChatId(evalCase, request, index);
            HybridRagAnswerService.HybridRagResult result = hybridRagAnswerService.answer(
                    evalCase.getQuestionText(),
                    tenant,
                    chatId,
                    ConversationIdHelper.build("eval", chatId),
                    request == null ? null : request.getModelProfile()
            );
            answer = emptyIfBlank(result.getAnswer());
            citations = toCitationStrings(result.getCitations());
            evidence = toEvidenceStrings(result.getEvidence());
            if (!StringUtils.hasText(answer)) {
                status = "FAILED";
                errorMessage = "empty answer";
            }
        } catch (RuntimeException ex) {
            status = "FAILED";
            errorMessage = StringUtils.hasText(ex.getMessage()) ? ex.getMessage() : "evaluation case failed";
        }

        long latencyMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedNs);
        EvaluationScorer.CaseScores scores = evaluationScorer.scoreCase(
                evalCase,
                answer,
                citations,
                evidence,
                "FAILED".equals(status)
        );
        EvalResultRecord record = EvalResultRecord.builder()
                .resultId("eval-result-" + shortUuid())
                .runId(runId)
                .datasetId(datasetId)
                .caseId(evalCase.getCaseId())
                .tenantId(tenant)
                .status(status)
                .questionText(evalCase.getQuestionText())
                .answerText(answer)
                .citationsJson(writeJsonList(citations))
                .evidenceJson(writeJsonList(evidence))
                .retrievalHit(scores.retrievalHit())
                .citationCoverage(scores.citationCoverage())
                .keywordScore(scores.keywordScore())
                .answerFaithfulness(scores.answerFaithfulness())
                .score(scores.score())
                .latencyMs(latencyMs)
                .errorMessage(errorMessage)
                .createdAt(LocalDateTime.now())
                .build();
        evalResultMapper.insert(record);
        return record;
    }

    private EvalMetricSummaryVO summarize(List<EvalResultRecord> results) {
        int total = results.size();
        int passed = (int) results.stream()
                .filter(item -> "SUCCESS".equals(item.getStatus()))
                .filter(item -> item.getScore() != null && item.getScore() >= PASS_THRESHOLD)
                .count();
        double totalSafe = Math.max(1, total);
        return EvalMetricSummaryVO.builder()
                .totalCases(total)
                .passedCases(passed)
                .runScore(round(avg(results.stream().map(EvalResultRecord::getScore).toList())))
                .retrievalHitRate(round(avg(results.stream().map(EvalResultRecord::getRetrievalHit).toList())))
                .citationCoverageRate(round(avg(results.stream().map(EvalResultRecord::getCitationCoverage).toList())))
                .answerFaithfulnessScore(round(avg(results.stream().map(EvalResultRecord::getAnswerFaithfulness).toList())))
                .avgLatencyMs(round(avg(results.stream().map(item -> item.getLatencyMs() == null ? null : item.getLatencyMs().doubleValue()).toList())))
                .failureRate(round(results.stream().filter(item -> !"SUCCESS".equals(item.getStatus())).count() / totalSafe))
                .build();
    }

    private EvalDatasetRecord requireDataset(String tenant, String datasetId) {
        if (!StringUtils.hasText(datasetId)) {
            throw new IllegalArgumentException("dataset id is required");
        }
        EvalDatasetRecord dataset = evalDatasetMapper.findByTenantAndDatasetId(tenant, datasetId.trim());
        if (dataset == null) {
            throw new IllegalArgumentException("dataset not found");
        }
        return dataset;
    }

    private EvalRunRecord requireRun(String tenant, String runId) {
        if (!StringUtils.hasText(runId)) {
            throw new IllegalArgumentException("run id is required");
        }
        EvalRunRecord run = evalRunMapper.findByTenantAndRunId(tenant, runId.trim());
        if (run == null) {
            throw new IllegalArgumentException("run not found");
        }
        return run;
    }

    private EvalDatasetVO toDatasetVO(EvalDatasetRecord record, int caseCount) {
        return EvalDatasetVO.builder()
                .datasetId(record.getDatasetId())
                .tenantId(record.getTenantId())
                .name(record.getName())
                .description(record.getDescription())
                .baselineRunId(record.getBaselineRunId())
                .caseCount(caseCount)
                .createdAt(formatTime(record.getCreatedAt()))
                .updatedAt(formatTime(record.getUpdatedAt()))
                .build();
    }

    private EvalRunVO toRunVO(EvalRunRecord run, List<EvalResultRecord> results) {
        EvalMetricSummaryVO metrics = EvalMetricSummaryVO.builder()
                .totalCases(intOrZero(run.getTotalCases()))
                .passedCases(intOrZero(run.getPassedCases()))
                .runScore(valueOrZero(run.getRunScore()))
                .retrievalHitRate(valueOrZero(run.getRetrievalHitRate()))
                .citationCoverageRate(valueOrZero(run.getCitationCoverageRate()))
                .answerFaithfulnessScore(valueOrZero(run.getAnswerFaithfulnessScore()))
                .avgLatencyMs(valueOrZero(run.getAvgLatencyMs()))
                .failureRate(valueOrZero(run.getFailureRate()))
                .build();
        return EvalRunVO.builder()
                .runId(run.getRunId())
                .datasetId(run.getDatasetId())
                .tenantId(run.getTenantId())
                .status(run.getStatus())
                .modelProfile(run.getModelProfile())
                .metrics(metrics)
                .results(results == null ? List.of() : results.stream().map(this::toResultVO).toList())
                .errorMessage(run.getErrorMessage())
                .startedAt(formatTime(run.getStartedAt()))
                .finishedAt(formatTime(run.getFinishedAt()))
                .createdAt(formatTime(run.getCreatedAt()))
                .build();
    }

    private EvalResultVO toResultVO(EvalResultRecord record) {
        return EvalResultVO.builder()
                .resultId(record.getResultId())
                .caseId(record.getCaseId())
                .status(record.getStatus())
                .question(record.getQuestionText())
                .answer(record.getAnswerText())
                .citations(readJsonList(record.getCitationsJson()))
                .evidence(readJsonList(record.getEvidenceJson()))
                .retrievalHit(valueOrZero(record.getRetrievalHit()))
                .citationCoverage(valueOrZero(record.getCitationCoverage()))
                .keywordScore(valueOrZero(record.getKeywordScore()))
                .answerFaithfulness(valueOrZero(record.getAnswerFaithfulness()))
                .score(valueOrZero(record.getScore()))
                .latencyMs(record.getLatencyMs() == null ? 0 : record.getLatencyMs())
                .errorMessage(record.getErrorMessage())
                .build();
    }

    private String resolveChatId(EvalCaseRecord evalCase, EvalRunRequestVO request, int index) {
        if (StringUtils.hasText(evalCase.getChatId())) {
            return evalCase.getChatId().trim();
        }
        if (request != null && StringUtils.hasText(request.getChatIdPrefix())) {
            return request.getChatIdPrefix().trim() + "-" + String.format(Locale.ROOT, "%03d", index + 1);
        }
        return evalCase.getDatasetId();
    }

    private List<String> toCitationStrings(List<CitationItem> items) {
        if (items == null || items.isEmpty()) {
            return List.of();
        }
        Set<String> values = new LinkedHashSet<>();
        for (CitationItem item : items) {
            if (item == null) {
                continue;
            }
            String text = "%s:%s:%s".formatted(
                    emptyIfBlank(item.getSourceType()),
                    emptyIfBlank(item.getTitle()),
                    emptyIfBlank(item.getChunkId())
            );
            if (StringUtils.hasText(text.replace(":", ""))) {
                values.add(text);
            }
        }
        return List.copyOf(values);
    }

    private List<String> toEvidenceStrings(List<EvidenceItem> items) {
        if (items == null || items.isEmpty()) {
            return List.of();
        }
        List<String> values = new ArrayList<>();
        for (EvidenceItem item : items) {
            if (item != null && StringUtils.hasText(item.getSnippet())) {
                values.add(item.getSnippet());
            }
        }
        return values;
    }

    private double avg(List<Double> values) {
        if (values == null || values.isEmpty()) {
            return 0.0;
        }
        double sum = 0.0;
        int count = 0;
        for (Double value : values) {
            if (value == null) {
                continue;
            }
            sum += value;
            count++;
        }
        return count == 0 ? 0.0 : sum / count;
    }

    private String writeJsonList(List<String> values) {
        try {
            return objectMapper.writeValueAsString(values == null ? List.of() : values);
        } catch (JsonProcessingException ex) {
            return "[]";
        }
    }

    private List<String> readJsonList(String json) {
        if (!StringUtils.hasText(json)) {
            return List.of();
        }
        try {
            return objectMapper.readValue(json, new TypeReference<List<String>>() {
            });
        } catch (JsonProcessingException ex) {
            return List.of();
        }
    }

    private double valueOrZero(Double value) {
        return value == null ? 0.0 : value;
    }

    private int intOrZero(Integer value) {
        return value == null ? 0 : value;
    }

    private double round(double value) {
        return Math.round(value * 10000.0) / 10000.0;
    }

    private String formatTime(LocalDateTime value) {
        return value == null ? "" : value.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
    }

    private String shortUuid() {
        return UUID.randomUUID().toString().replace("-", "").substring(0, 12);
    }

    private String emptyIfBlank(String value) {
        return StringUtils.hasText(value) ? value : "";
    }
}

File: typescript/apps/api/src/evaluation/evaluation.service.ts
import { randomBytes } from "node:crypto";

import { BadRequestException, Injectable } from "@nestjs/common";
import type { Citation } from "@knowledgeops/shared";

import { AiService } from "../ai/ai.service.js";
import { normalizeTenant } from "../common/tenant.js";
import { type EvalDataset, type EvalRun, PlatformStore } from "../platform/platform.store.js";
import { EvaluationReportRenderer } from "./evaluation-report.renderer.js";
import { EvaluationScorer, roundEvaluation } from "./evaluation.scorer.js";

const PASS_THRESHOLD = 0.70;

export interface EvaluationCasePayload {
  caseId?: string;
  category?: string;
  chatId?: string;
  question?: string;
  expectedCitations?: string[];
  expectedKeywords?: string[];
  forbiddenKeywords?: string[];
}

export interface EvaluationDatasetPayload {
  name?: string;
  description?: string;
  cases?: Array<EvaluationCasePayload | null>;
}

export interface EvaluationRunRequest {
  datasetId?: string;
  modelProfile?: string;
  chatIdPrefix?: string;
}

export interface EvaluationDatasetView {
  datasetId: string;
  tenantId: string;
  name: string;
  description: string;
  baselineRunId: string | null;
  caseCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface EvaluationMetrics {
  [key: string]: number;
  totalCases: number;
  passedCases: number;
  runScore: number;
  retrievalHitRate: number;
  citationCoverageRate: number;
  answerFaithfulnessScore: number;
  avgLatencyMs: number;
  failureRate: number;
}

export interface EvaluationResultView {
  resultId: string;
  caseId: string;
  status: string;
  question: string;
  answer: string;
  citations: string[];
  evidence: string[];
  retrievalHit: number;
  citationCoverage: number;
  keywordScore: number;
  answerFaithfulness: number;
  score: number;
  latencyMs: number;
  errorMessage: string;
}

export interface EvaluationRunView {
  runId: string;
  datasetId: string;
  tenantId: string;
  status: string;
  modelProfile: string;
  metrics: EvaluationMetrics;
  results: EvaluationResultView[];
  errorMessage: string | null;
  startedAt: string;
  finishedAt: string;
  createdAt: string;
}

export interface EvaluationComparisonView {
  dataset: EvaluationDatasetView;
  baseline: EvaluationRunView | null;
  current: EvaluationRunView | null;
}

@Injectable()
export class EvaluationService {
  constructor(
    private readonly store: PlatformStore,
    private readonly aiService: AiService,
    private readonly scorer: EvaluationScorer,
    private readonly reportRenderer: EvaluationReportRenderer
  ) {}

  createDataset(tenantId: string, request: EvaluationDatasetPayload | null | undefined): EvaluationDatasetView {
    if (!request) {
      throw new BadRequestException("dataset payload is required");
    }
    if (!request.name?.trim()) {
      throw new BadRequestException("dataset name is required");
    }
    if (!request.cases?.length) {
      throw new BadRequestException("dataset cases are required");
    }

    const cases = request.cases.map((item, index) => normalizeCase(item, index));
    const tenant = normalizeTenant(tenantId);
    const now = localDateTime();
    const dataset: EvalDataset = {
      datasetId: evaluationId("eval-ds"),
      tenantId: tenant,
      name: request.name.trim(),
      description: nonBlankOrEmpty(request.description),
      cases,
      createdAt: now,
      updatedAt: now
    };
    this.store.evalDatasets.set(dataset.datasetId, dataset);
    this.store.persist();
    return datasetView(dataset);
  }

  listDatasets(tenantId: string): EvaluationDatasetView[] {
    const tenant = normalizeTenant(tenantId);
    return [...this.store.evalDatasets.values()]
      .filter((dataset) => dataset.tenantId === tenant)
      .sort((left, right) => right.updatedAt.localeCompare(left.updatedAt))
      .map(datasetView);
  }

  async triggerRun(
    tenantId: string,
    datasetId: string,
    request?: EvaluationRunRequest | null
  ): Promise<EvaluationRunView> {
    const tenant = normalizeTenant(tenantId);
    const dataset = this.requireDataset(tenant, datasetId);
    if (dataset.cases.length === 0) {
      throw new BadRequestException("dataset has no cases");
    }

    const now = localDateTime();
    const modelProfile = request?.modelProfile?.trim() || "balanced";
    const run: EvalRun = {
      runId: evaluationId("eval-run"),
      datasetId: dataset.datasetId,
      tenantId: tenant,
      status: "RUNNING",
      modelProfile,
      metrics: emptyMetrics(dataset.cases.length),
      results: [],
      createdAt: now,
      startedAt: now
    };
    this.store.evalRuns.set(run.runId, run);
    this.store.persist();

    const results: EvaluationResultView[] = [];
    for (const [index, evalCase] of dataset.cases.entries()) {
      results.push(await this.runCase(tenant, run.runId, dataset, evalCase, request, index));
    }

    run.status = "SUCCESS";
    run.metrics = summarize(results);
    run.results = results.map((result) => ({ ...result }));
    run.finishedAt = localDateTime();
    this.store.persist();
    return runView(run);
  }

  getRun(tenantId: string, runId: string): EvaluationRunView {
    return runView(this.requireRun(normalizeTenant(tenantId), runId));
  }

  markBaseline(tenantId: string, runId: string): EvaluationRunView {
    const tenant = normalizeTenant(tenantId);
    const run = this.requireRun(tenant, runId);
    const dataset = this.store.evalDatasets.get(run.datasetId);
    if (!dataset || dataset.tenantId !== tenant) {
      throw new BadRequestException("dataset not found");
    }
    dataset.baselineRunId = run.runId;
    dataset.updatedAt = localDateTime();
    this.store.persist();
    return runView(run);
  }

  compareLatest(tenantId: string, datasetId: string): EvaluationComparisonView {
    const tenant = normalizeTenant(tenantId);
    const dataset = this.requireDataset(tenant, datasetId);
    const recent = [...this.store.evalRuns.values()]
      .filter((run) => run.tenantId === tenant && run.datasetId === dataset.datasetId)
      .map((run, index) => ({ run, index }))
      .sort((left, right) => right.run.createdAt.localeCompare(left.run.createdAt) || right.index - left.index)
      .slice(0, 2)
      .map(({ run }) => run);
    const current = recent[0];

    let baseline: EvalRun | undefined;
    if (dataset.baselineRunId) {
      const candidate = this.store.evalRuns.get(dataset.baselineRunId);
      if (candidate?.tenantId === tenant) {
        baseline = candidate;
      }
    }
    baseline ??= recent[1];

    return {
      dataset: datasetView(dataset),
      baseline: baseline ? runView(baseline) : null,
      current: current ? runView(current) : null
    };
  }

  exportReport(tenantId: string, runId: string): string {
    return this.reportRenderer.render(this.getRun(tenantId, runId));
  }

  private async runCase(
    tenant: string,
    runId: string,
    dataset: EvalDataset,
    evalCase: Record<string, unknown>,
    request: EvaluationRunRequest | null | undefined,
    index: number
  ): Promise<EvaluationResultView> {
    const started = process.hrtime.bigint();
    let status = "SUCCESS";
    let answer = "";
    let errorMessage = "";
    let citations: string[] = [];
    let evidence: string[] = [];

    try {
      const chatId = resolveChatId(dataset.datasetId, evalCase, request, index);
      const result = await this.aiService.reactChat({
        prompt: String(evalCase.question ?? ""),
        chatId,
        modelProfile: request?.modelProfile
      }, tenant, "eval");
      answer = nonBlankOrEmpty(result.answer);
      citations = citationStrings(result.citations ?? []);
      evidence = stringList(result.evidence).filter((item) => item.trim());
      if (!answer.trim()) {
        status = "FAILED";
        errorMessage = "empty answer";
      }
    } catch (error) {
      status = "FAILED";
      errorMessage = error instanceof Error && error.message.trim()
        ? error.message
        : "evaluation case failed";
    }

    const latencyMs = Number((process.hrtime.bigint() - started) / 1_000_000n);
    const scores = this.scorer.scoreCase(evalCase, answer, citations, evidence, status === "FAILED");
    return {
      resultId: evaluationId("eval-result"),
      caseId: String(evalCase.caseId ?? ""),
      status,
      question: String(evalCase.question ?? ""),
      answer,
      citations,
      evidence,
      ...scores,
      latencyMs,
      errorMessage
    };
  }

  private requireDataset(tenant: string, datasetId: string): EvalDataset {
    if (!datasetId?.trim()) {
      throw new BadRequestException("dataset id is required");
    }
    const dataset = this.store.evalDatasets.get(datasetId.trim());
    if (!dataset || dataset.tenantId !== tenant) {
      throw new BadRequestException("dataset not found");
    }
    return dataset;
  }

  private requireRun(tenant: string, runId: string): EvalRun {
    if (!runId?.trim()) {
      throw new BadRequestException("run id is required");
    }
    const run = this.store.evalRuns.get(runId.trim());
    if (!run || run.tenantId !== tenant) {
      throw new BadRequestException("run not found");
    }
    return run;
  }
}

function normalizeCase(item: EvaluationCasePayload | null, index: number): Record<string, unknown> {
  if (!item?.question?.trim()) {
    throw new BadRequestException("case question is required");
  }
  return {
    caseId: item.caseId?.trim() || `case-${String(index + 1).padStart(3, "0")}`,
    category: nonBlankOrEmpty(item.category),
    chatId: nonBlankOrEmpty(item.chatId),
    question: item.question.trim(),
    expectedCitations: stringList(item.expectedCitations),
    expectedKeywords: stringList(item.expectedKeywords),
    forbiddenKeywords: stringList(item.forbiddenKeywords)
  };
}

function resolveChatId(
  datasetId: string,
  evalCase: Record<string, unknown>,
  request: EvaluationRunRequest | null | undefined,
  index: number
): string {
  const caseChatId = typeof evalCase.chatId === "string" ? evalCase.chatId.trim() : "";
  if (caseChatId) {
    return caseChatId;
  }
  const prefix = request?.chatIdPrefix?.trim();
  return prefix ? `${prefix}-${String(index + 1).padStart(3, "0")}` : datasetId;
}

function citationStrings(citations: Citation[]): string[] {
  const values = new Set<string>();
  for (const citation of citations) {
    if (!citation) {
      continue;
    }
    const sourceType = nonBlankOrEmpty((citation as Citation & { sourceType?: string }).sourceType ?? citation.source);
    const title = nonBlankOrEmpty(citation.title);
    const chunkId = nonBlankOrEmpty(citation.chunkId);
    const text = `${sourceType}:${title}:${chunkId}`;
    if (text.replaceAll(":", "").trim()) {
      values.add(text);
    }
  }
  return [...values];
}

function summarize(results: EvaluationResultView[]): EvaluationMetrics {
  const totalCases = results.length;
  const passedCases = results
    .filter((result) => result.status === "SUCCESS" && result.score >= PASS_THRESHOLD)
    .length;
  const denominator = Math.max(1, totalCases);
  return {
    totalCases,
    passedCases,
    runScore: roundEvaluation(average(results.map((result) => result.score))),
    retrievalHitRate: roundEvaluation(average(results.map((result) => result.retrievalHit))),
    citationCoverageRate: roundEvaluation(average(results.map((result) => result.citationCoverage))),
    answerFaithfulnessScore: roundEvaluation(average(results.map((result) => result.answerFaithfulness))),
    avgLatencyMs: roundEvaluation(average(results.map((result) => result.latencyMs))),
    failureRate: roundEvaluation(results.filter((result) => result.status !== "SUCCESS").length / denominator)
  };
}

function emptyMetrics(totalCases: number): EvaluationMetrics {
  return {
    totalCases,
    passedCases: 0,
    runScore: 0,
    retrievalHitRate: 0,
    citationCoverageRate: 0,
    answerFaithfulnessScore: 0,
    avgLatencyMs: 0,
    failureRate: 0
  };
}

function average(values: number[]): number {
  return values.length === 0 ? 0 : values.reduce((sum, value) => sum + value, 0) / values.length;
}

function datasetView(dataset: EvalDataset): EvaluationDatasetView {
  return {
    datasetId: dataset.datasetId,
    tenantId: dataset.tenantId,
    name: dataset.name,
    description: dataset.description ?? "",
    baselineRunId: dataset.baselineRunId ?? null,
    caseCount: dataset.cases.length,
    createdAt: dataset.createdAt,
    updatedAt: dataset.updatedAt
  };
}

function runView(run: EvalRun): EvaluationRunView {
  return {
    runId: run.runId,
    datasetId: run.datasetId,
    tenantId: run.tenantId,
    status: run.status,
    modelProfile: run.modelProfile,
    metrics: metricsView(run.metrics),
    results: run.results.map(resultView),
    errorMessage: run.errorMessage ?? null,
    startedAt: run.startedAt ?? "",
    finishedAt: run.finishedAt ?? "",
    createdAt: run.createdAt
  };
}

function metricsView(metrics: Record<string, number>): EvaluationMetrics {
  return {
    totalCases: numberOrZero(metrics.totalCases),
    passedCases: numberOrZero(metrics.passedCases),
    runScore: numberOrZero(metrics.runScore),
    retrievalHitRate: numberOrZero(metrics.retrievalHitRate),
    citationCoverageRate: numberOrZero(metrics.citationCoverageRate),
    answerFaithfulnessScore: numberOrZero(metrics.answerFaithfulnessScore),
    avgLatencyMs: numberOrZero(metrics.avgLatencyMs),
    failureRate: numberOrZero(metrics.failureRate)
  };
}

function resultView(result: Record<string, unknown>): EvaluationResultView {
  return {
    resultId: String(result.resultId ?? ""),
    caseId: String(result.caseId ?? ""),
    status: String(result.status ?? ""),
    question: String(result.question ?? ""),
    answer: String(result.answer ?? ""),
    citations: stringList(result.citations),
    evidence: stringList(result.evidence),
    retrievalHit: numberOrZero(result.retrievalHit),
    citationCoverage: numberOrZero(result.citationCoverage),
    keywordScore: numberOrZero(result.keywordScore),
    answerFaithfulness: numberOrZero(result.answerFaithfulness),
    score: numberOrZero(result.score),
    latencyMs: numberOrZero(result.latencyMs),
    errorMessage: String(result.errorMessage ?? "")
  };
}

function stringList(value: unknown): string[] {
  return Array.isArray(value)
    ? value.map((item) => typeof item === "string" ? item : "")
    : [];
}

function numberOrZero(value: unknown): number {
  const number = Number(value);
  return Number.isFinite(number) ? number : 0;
}

function nonBlankOrEmpty(value: unknown): string {
  return typeof value === "string" && value.trim() ? value : "";
}

function evaluationId(prefix: string): string {
  return `${prefix}-${randomBytes(6).toString("hex")}`;
}

function localDateTime(date = new Date()): string {
  const offsetMs = date.getTimezoneOffset() * 60_000;
  return new Date(date.getTime() - offsetMs).toISOString().replace(/Z$/, "");
}
